### DISCUSSION TITLE
Search function not working properly?

### DISCUSSION THREAD

**Discussion by sotono-046** [xesscorp/skidl#183] [Q&A]
Hi,There.

I am trying to use SKiDL for the first time for the first time and I am touching it with the documentation.
I stumbled on the part about finding the parts list at the beginning.

[SKiDL — SKiDL](https://devbisme.github.io/skidl/#finding-parts)

```python
from skidl import *

def main():
    search("opamp")

if __name__ == "__main__":
    main()

```

After doing this, you should see a list of opamps.
However, the actual output was as follows.

```
Device.kicad_sym: Opamp_Dual ( )
Device.kicad_sym: Opamp_Quad ( )
Simulation_SPICE.kicad_sym: OPAMP (
```

The latest version of KiCad is of course installed and environment variables are already set.

Why is the list of parts not displaying correctly?

**Comment by ckirchhefer** [comment]
Funny thing I looked into this forum for the exact same reason just now. Also using SKiDL for the first time and struggling with the search operations. 
I would like to add that I'm currently not able to get any results from the `search_footprints` function.

**Comment by sotono-046** [comment]
I have since tried it.

## I tried version 6.
I tried changing the KiCad version from 7 to 6, but there was no change.
So I am using version 7.

## Exported parts from KiCad
I exported the parts from KiCad and combined them into one symbol, and then I tried to incorporate it into a Python environment variable, and I was able to successfully load the parts for that symbol.
However, there was nothing in the parentheses as well.
``
The following is the code.
```python
from skidl import *

set_default_tool(KICAD)

symbol_path = "hardware"
lib_search_paths[KICAD].append(symbol_path)

footprint_path = "hardware/Switch_Keyboard_Hotswap_Kailh.pretty"
footprint_search_paths[KICAD].append(footprint_path)

search("1n4148")
```


I think it is probably usable just because I can't find the part, but I haven't tried that far!
Sorry.

**Reply by 2e0byo** [reply]
Hum, I think the parsing of v6 libraries is broken, probably involving inheritance.  Try version kicad libraries v. 5 (which has a completely different syntax). [edit: inherited/extended parts, rather than python class inheritance]

**Comment by sotono-046** [comment]
I am trying to test that now, but I have another error so I cannot verify it.
```python
<urlopen error unknown url type: c>
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 11, in search_parts
    skidl.search(part_name)
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 15, in main
    search_parts()
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 19, in <module>
    main()
urllib.error.URLError: <urlopen error unknown url type: c>
```

I had to set it up again because of the new environment, but when I run `skidl.search('diode')` I get this error.
Is there any promising solution?

**Reply by 2e0byo** [reply]
urlopen?!  didn't realise skidl accesses the internet.

What's the full traceback?  the immediate error is that something is trying to open a url of the form `c://something`, but goodness knows why.

**Reply by sotono-046** [reply]
OK, I tried to execute this code to search for parts.

```python
import skidl

def search_parts():
    part_names = ["diode", "switch", "RP2040", "oled"]
    for part_name in part_names:
        print(f"---------- {part_name}")
        skidl.search(part_name)

def main():
    search_parts()

if __name__ == "__main__":
    main()
```

And when I run this, I get this error in `skidl.seartch(part_name)`.
By the way, Sotono is my user name and another name.

```python
~\Desktop\Python_diy_keyboard via 🐍 v3.11.4 (.venv) 
⚡ ➜  & 'c:\Users\Sotono\Desktop\Python_diy_keyboard\.venv\Scripts\python.exe' 'C:\Users\Sotono\scoop\persist\vscode\data\extensions\ms-python.python-2023.12.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher' '55230' '--' 'C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py'
---------- diode
Traceback (most recent call last):
  File "C:\Users\Sotono\scoop\apps\python\current\Lib\runpy.py", line 198, in _run_module_as_main
    return _run_code(code, main_globals, None,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sotono\scoop\apps\python\current\Lib\runpy.py", line 88, in _run_code
    exec(code, run_globals)
  File "C:\Users\Sotono\scoop\persist\vscode\data\extensions\ms-python.python-2023.12.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher/../..\debugpy\__main__.py", line 39, in <module>    
    cli.main()
  File "C:\Users\Sotono\scoop\persist\vscode\data\extensions\ms-python.python-2023.12.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher/../..\debugpy/..\debugpy\server\cli.py", line 430, in main
    run()
  File "C:\Users\Sotono\scoop\persist\vscode\data\extensions\ms-python.python-2023.12.0\pythonFiles\lib\python\debugpy\adapter/../..\debugpy\launcher/../..\debugpy/..\debugpy\server\cli.py", line 284, in run_file
    runpy.run_path(target, run_name="__main__")
  File "C:\Users\Sotono\scoop\persist\vscode\data\extensions\ms-python.python-2023.12.0\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_runpy.py", line 321, in run_path
    return _run_module_code(code, init_globals, run_name,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sotono\scoop\persist\vscode\data\extensions\ms-python.python-2023.12.0\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_runpy.py", line 135, in _run_module_code    
    _run_code(code, mod_globals, init_globals,
  File "C:\Users\Sotono\scoop\persist\vscode\data\extensions\ms-python.python-2023.12.0\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydevd_bundle\pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 19, in <module>
    main()
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 15, in main
    search_parts()
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 11, in search_parts
    skidl.search(part_name)
  File "c:\Users\Sotono\Desktop\Python_diy_keyboard\.venv\Lib\site-packages\skidl\part_query.py", line 131, in search_parts
    for part in search_parts_iter(terms, tool):
  File "c:\Users\Sotono\Desktop\Python_diy_keyboard\.venv\Lib\site-packages\skidl\part_query.py", line 107, in search_parts_iter
    lib = SchLib(
          ^^^^^^^
  File "c:\Users\Sotono\Desktop\Python_diy_keyboard\.venv\Lib\site-packages\skidl\schlib.py", line 88, in __init__
    tool_modules[tool].load_sch_lib(
  File "c:\Users\Sotono\Desktop\Python_diy_keyboard\.venv\Lib\site-packages\skidl\tools\kicad\kicad.py", line 80, in load_sch_lib
    f, _ = find_and_open_file(
           ^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Sotono\Desktop\Python_diy_keyboard\.venv\Lib\site-packages\skidl\utilities.py", line 793, in find_and_open_file
    return urllib.request.urlopen(link), link
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sotono\scoop\apps\python\current\Lib\urllib\request.py", line 216, in urlopen       
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sotono\scoop\apps\python\current\Lib\urllib\request.py", line 519, in open
    response = self._open(req, data)
               ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sotono\scoop\apps\python\current\Lib\urllib\request.py", line 541, in _open
    return self._call_chain(self.handle_open, 'unknown',
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Sotono\scoop\apps\python\current\Lib\urllib\request.py", line 496, in _call_chain   
    result = func(*args)
             ^^^^^^^^^^^
  File "C:\Users\Sotono\scoop\apps\python\current\Lib\urllib\request.py", line 1419, in unknown_open 
    raise URLError('unknown url type: %s' % type)
urllib.error.URLError: <urlopen error unknown url type: c>
```

Now that I am looking at this traceback, it seems to have something to do with Scoop, so I will investigate Scoop after I post this!

**Reply by 2e0byo** [reply]
Total guess but you probably have `c:/` rather than `r"c:\path` somewhere.  windows paths are a pain.  this would cause something to think it had a url with the scheme `c`, hence the error.

I don't have windows and don't know if skidl works on it, but someone who does will probably be along shortly.

**Reply by 2e0byo** [reply]
If you haven't set a path anywhere, install `rich` and add 

```python
from rich.traceback import install
install(show_locals=True)
```

To the top of your script.  Not only will the output be much nicer, it will show us what that path is and where it's coming from.

**Comment by sotono-046** [comment]
I thought there was something wrong with the Python installed by Scoop, so I did a clean install, but the situation did not change.
I changed the Python version and the exact same error occurred.

As a test, I rewrote the path pointing to the C drive as the Z drive at random, and there was progress.

```python
WARNING: Could not open directory 'Z:/Program Files/KiCad/share/kicad/library' @ [C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py:18=>C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py:14]
```

This proves that the path I set up is correct.

I got that error because of the path, but I remember it working in a Windows environment when I touched it before.
Therefore, I cannot give up working under Windows environment.

Tomorrow I will wake up in the morning and I will be in an environment other than Scoop (e.g. winget?). I will try to run Python in a non-Scoop environment (e.g., winget?) when I wake up in the morning.
I took a closer look at my environment and found that I had an old Python (3.10) installed. (But I uninstalled it, because it was confusing me!)
Perhaps something not so good is happening in Scoop's environment? I can't deny the possibility that this is not a good thing in Scoop's environment.

I will write again tomorrow if I find out something.
Thanks again.

**Reply by 2e0byo** [reply]
'Z:/Program Files/KiCad/share/kicad/library' this path is wrong.  It should be `r"Z:\Program Files\KiCad` etc.  Windows, for bug-for-bug compatibility with a brain-dead decision in the past, uses `\` not `/` as a path separator.  Where does that line come from?

> this proves that the path I set up is correct

Not really, I think it just shows that for whatever reason `c:/` looks like a url (maybe you had `c://`?, but `Z:/` doesn't.  But neither is correct.

Could you post the code where you set that path?

**Reply by sotono-046** [reply]
I use the `.env` file because I don't want to mess up my environment variables too much (which is terribly selfish) and I know Python reads this.
By the way, the `.env` looks like this

```
KICAD_SYMBOL_DIR=C:/Program Files/KiCad/share/kicad/library
```

I was verifying the path by changing the `c` at the beginning of this to `z`, but this was misleading. The `c` is correct.

As 2e0byo points out, windows uses `\` for path notation, but I know that realh works in Python by changing it to `/`.
This comes down to preference, but I don't like having to type `\` twice because it could be escaped, so I use `/`.

…Incidentally, I switched package managers from Scoop to Winget with the same results.

Is it still due to the fact that it is Windows? (However, I am sure I remember that it used to work before...)

**Reply by sotono-046** [reply]
Once organized.

I did the following to clean up my environment
- Delete the `.venv` folder, which is the virtual environment
- Run `python -m venv .venv`
- Run `pip install skidl`
- Write `KICAD_SYMBOL_DIR=C:\Program Files\KiCad\share\kicad\library` in `.env`.
- Activate and run `.venv

And the error that comes up is below.
```py
Exception occurred: URLError
<urlopen error unknown url type: c>
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 14, in search_parts
    skidl.search(part_name)
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 18, in main
    search_parts()
  File "C:\Users\Sotono\Desktop\Python_diy_keyboard\main.py", line 22, in <module
    main()
urllib.error.URLError: <urlopen error unknown url type: c>
```

My environment is as follows
- Windows 11
- Python 3.11.4
- Latest version of Skidl

**Reply by 2e0byo** [reply]
Oh you're right it doesn't matter.  I'm not a windows dev and I didn't know the paths are automagically fixed.  But [this fn](https://github.com/devbisme/skidl/blob/9f49c38a5ab6d5fb4f2a7a09edeca3af2a910118/src/skidl/utilities.py#L757-L758) is broken anyway and never going to work on windows.  Hang on I'll make a PR fixing it.

**Reply by sotono-046** [reply]
Thank you.
In the meantime I will try to run Linux on WSL and see if I can run skidl there. (I think it will work)

**Reply by 2e0byo** [reply]
https://github.com/devbisme/skidl/pull/189

You can try this by cloning the branch.  There might be other issues, but this should fix the `is_url` function.

**Comment by sotono-046** [comment]
OK, I tried.
```
pip install git+https://github.com/devbisme/skidl.git@development
```

And it ran fine, no errors.
Thank you.
Now we are going to make a good one!

**Comment by devbisme** [comment]
OK, good. Thanks for the reports and the fix.
